### DISCUSSION TITLE
Broadcasts should have code generation to make them simpler like RPCs

### DISCUSSION THREAD

**Discussion by Pfhoenix** [FirstGearGames/FishNet#622] [Feature Request]
Currently, broadcasts require you to define a struct that extends IBroadcast, every method that handles the broadcast has to have a set signature (clients have [broadcasttype, channel] while servers have [networkconnection, broadcasttype, channel]), and the developer has to manually register methods to handle broadcasts by the broadcasttype.

This is a lot of manual work for something that can be done entirely via code-generation. Given the following methods (for the sake of an example) :
```
[ClientBroadcast]
ClientReceiveCount(int count)
```
- similar to RPCs, have a ServerBroadcastAttribute and ClientBroadcastAttribute
- any method with either attribute gets a new type generated specifically for it, so the above method would have a struct ClientReceiveCountDataBroadcast : IBroadcast { public int count; } generated
- a wrapper method then generated, ClientReceiveCount_Wrapper(ClientReceiveCountDataBroadcast broadcastData, Channel channel) , which calls ClientReceiveCount(broadcastData.count), and the wrapper method would immediately be registered with ClientManager
- any calls to ClientReceiveCount then get swapped with InstanceFinder.ServerManager.Broadcast, generating the requisite broadcasttype with some simple field-by-name matching via reflection

Now, I'm not entirely sure what the best way to do a broadcast to a single client would be, but the above is very straightforwards considering what Fish-Net already does for RPCs. Now, why would someone go through all that trouble instead of using RPCs? Simple - can't use RPCs when server and clients are in different scenes, or they want objects to be able to call on each other without requiring the objects be in sync between server and client.

**Comment by FirstGearGames** [comment]
Broadcasts are consider low level and exist outside networking scripts. Codegen uses attributes and other things from networked scripts to create automatic handling of things.

The intended behavior of broadcasts is as they are so I'm going to have to decline this. Let me know if you have any other questions about this.

**Reply by Pfhoenix** [reply]
I understand that the current implementation is as intended. I'm not sure why codegen is being limited to networked scripts when the same codegen could make life using broadcasts much simpler. What would you charge for this proposed feature to be implemented? If it's not something you're willing to consider, could you point me in the right direction to implement it myself, then?

**Reply by Monsoonexe** [reply]
Hi @Pfhoenix!
It's possible to modify the weaver yourself to generate whatever code you want. Code gen is a strange beast, but, like most things with programming, gets really easy once you know what's going on.

If you have any questions about weaving/code gen, feel free to hit me up here or on Discord (monsoonexe).
